third_party/pcsc-lite/naclport/server_clients_management/src/client_request_processor.cc

```diff
@@ -410,33 +437,118 @@ GenericRequestResult PcscLiteClientRequestProcessor::SCardConnect(
   tracer.LogEntrance();
 
   LONG return_code = SCARD_S_SUCCESS;
-  if (!s_card_handles_registry_.ContainsContext(s_card_context))
+  if (update_card_handles_registry &&
+      !s_card_handles_registry_.ContainsContext(s_card_context)) {
     return_code = SCARD_E_INVALID_HANDLE;
+  }
 
-  SCARDHANDLE s_card_handle;
-  DWORD active_protocol;
   if (return_code == SCARD_S_SUCCESS) {
     return_code =
         ::SCardConnect(s_card_context, reader_name.c_str(), share_mode,
-                       preferred_protocols, &s_card_handle, &active_protocol);
+                       preferred_protocols, s_card_handle, active_protocol);
 
     // Catch when PC/SC-Lite core, unlike us, thinks the context doesn't exist.
     if (return_code == SCARD_E_INVALID_HANDLE)
       OnSCardContextRevoked(s_card_context);
   }
 
+  if (return_code == SCARD_E_PROTO_MISMATCH && handle_protocol_mismatch_error) {
+    return_code = SCardConnectHandleProtocolMismatchError(
+        s_card_context, reader_name, share_mode, preferred_protocols,
+        s_card_handle, active_protocol);
+  }
+
   tracer.AddReturnValue(DebugDumpSCardReturnCode(return_code));
   if (return_code == SCARD_S_SUCCESS) {
-    tracer.AddReturnedArg("hCard", DebugDumpSCardHandle(s_card_handle));
+    tracer.AddReturnedArg("hCard", DebugDumpSCardHandle(*s_card_handle));
     tracer.AddReturnedArg("dwActiveProtocol",
-                          DebugDumpSCardProtocol(active_protocol));
+                          DebugDumpSCardProtocol(*active_protocol));
   }
   tracer.LogExit();
 
-  if (return_code != SCARD_S_SUCCESS)
-    return ReturnValues(return_code);
-  s_card_handles_registry_.AddHandle(s_card_context, s_card_handle);
-  return ReturnValues(return_code, s_card_handle, active_protocol);
+  if (return_code == SCARD_S_SUCCESS && update_card_handles_registry)
+    s_card_handles_registry_.AddHandle(s_card_context, *s_card_handle);
+
+  return return_code;
+}
+
+LONG PcscLiteClientRequestProcessor::SCardConnectHandleProtocolMismatchError(
+    SCARDCONTEXT s_card_context,
+    const std::string& reader_name,
+    DWORD share_mode,
+    DWORD preferred_protocols,
+    LPSCARDHANDLE s_card_handle,
+    LPDWORD active_protocol) {
+  LONG return_code = SCARD_E_PROTO_MISMATCH;
+  // If SCardConnect fails with SCARD_E_PROTO_MISMATCH it could be because a
+  // client application did not correctly reset the card and is now trying to
+  // reuse the previous connection although it requires a different protocol. If
+  // allowed by admin policy, disconnect and reset the card, and retry
+  // connecting to it.
+
+  optional<AdminPolicy> admin_policy = admin_policy_getter_->WaitAndGet();
+  std::vector<std::string> client_ids;
+  if (admin_policy &&
+      admin_policy.value().scard_disconnect_fallback_client_app_ids) {
+    client_ids =
+        admin_policy.value().scard_disconnect_fallback_client_app_ids.value();
+  }
+
+  if (std::find(client_ids.begin(), client_ids.end(),
```

Comment:
nit: Let's simplify this code to reduce the nesting level. Also we don't need to declare `return_code` that early:

```
  if (std::find(...) == client_ids.end()) {
    // The fallback is not enabled via policy, so return the original error.
    return SCARD_E_PROTO_MISMATCH;
  }
  ...
```